toolchain/check/check_unit.cpp

```diff
@@ -570,10 +571,81 @@ auto CheckUnit::CheckPoisonedConcreteImplLookupQueries() -> void {
   context_.inst_block_stack().PopAndDiscard();
 }
 
+auto CheckUnit::CheckOverlappingImpls() -> void {
+  // Collect all of the impls sorted into contiguous segments by their
+  // interface. We only need to compare impls within each such segment.
+  llvm::SmallVector<SemIR::Impl> impls_by_interface(
+      context_.impls().array_ref());
+  llvm::sort(
+      impls_by_interface, [](const SemIR::Impl& a, const SemIR::Impl& b) {
+        return a.interface.interface_id.index < b.interface.interface_id.index;
+      });
+
+  const auto* it = impls_by_interface.begin();
+  while (it != impls_by_interface.end()) {
+    const auto* segment_begin = it;
+    do {
+      ++it;
+    } while (it != impls_by_interface.end() &&
+             it->interface.interface_id ==
+                 segment_begin->interface.interface_id);
+    const auto* segment_end = it;
+
+    if (std::distance(segment_begin, segment_end) == 1) {
+      // Only 1 interface in the segment; nothing to overlap with.
+      continue;
+    }
+
+    CheckOverappingImplsForInterface(
+        llvm::ArrayRef(segment_begin, segment_end));
+  }
+}
+
+auto CheckUnit::CheckOverappingImplsForInterface(
+    llvm::ArrayRef<SemIR::Impl> impls) -> void {
+  for (auto [index, impl] : llvm::enumerate(impls)) {
+    if (impl.witness_id == SemIR::ErrorInst::InstId) {
+      continue;
+    }
+    auto type_structure =
+        BuildTypeStructure(context_, impl.self_id, impl.interface);
+
+    for (const auto& impl2 : llvm::drop_begin(impls, index + 1)) {
+      if (impl2.witness_id == SemIR::ErrorInst::InstId) {
+        continue;
+      }
+
+      // The type structure each non-final `impl` must differ from all other
+      // non-final `impl` for the same interface visible from the file.
+      if (!impl.is_final && !impl2.is_final) {
+        auto type_structure2 =
+            BuildTypeStructure(context_, impl2.self_id, impl2.interface);
+        if (type_structure == type_structure2) {
+          CARBON_DIAGNOSTIC(
+              ImplOverlapNonFinal, Error,
+              "found non-final `impl` with identical type structure");
+          auto builder =
+              emitter_.Build(impl2.latest_decl_id(), ImplOverlapNonFinal);
+          CARBON_DIAGNOSTIC(ImplOverlapNonFinalNote, Note,
+                            "it overlaps with the `impl` here");
```

Comment:
I think you could drop words:
```suggestion
                            "overlaps with `impl` here");
```

And possibly also say "identical to", as in:
```suggestion
                            "identical to `impl` here");
```

Reply:
I was hedging here because I want to talk about overlap instead of type structure, which I think isn't a thing we want people to know to use the language maybe? But the result is unsatisfactory as you have called out here :) I don't actually know how to write this error without mentioning type structure right now, so going with your second suggestion here.

Reply:
FWIW, I'm not actually sure why "overlap" is the term being used here ([overlap rule](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#overlap-rule)?). But, I'd expect consistent discussion of whatever term is used, whether that's "overlap" or "identical". e.g., you could also instead do "identical type structure" -> "overlapping type structure"

Reply:
To be sure, if this is the overlap rule, I think we should feel okay saying "overlapping" and just making sure this is the only way we use "overlap" in the context

Reply:
Yeah, it is the overlap rule except to the point where the overlap rule can't pick one because the type structures are identical. So they are "fully overlapping" is how I thought of it. But just "overlapping" isn't enough because partial overlap is fine and we use the overlap rule to sort them.

Reply:
I wouldn't object to "fully overlaps with" / "fully overlapping", although maybe "identical" communicates that better

Reply:
To be clear, I don't have a strong opinion here past that the note should use the same term; either way, feel feel free to merge once tests are passing.